pkg/ipcache/types.go

```diff
@@ -176,5 +229,33 @@ func (s prefixInfo) logConflicts(scopedLog *logrus.Entry) {
 				overrideResourceID = resourceID
 			}
 		}
+
+		if info.tunnelPeer.IsValid() {
+			if tunnelPeer.IsValid() {
```

Comment:
won't this always evaluate true after the 1st iteration with a valid `info.tunnelPeer`? or is the expectation here that only a single `info.tunnelPeer.IsValid()` in the given range?

Reply:
cc @gandro

Reply:
Yes, there is the expectation that there is only a single info.tunnelPeer.IsValid() in the given range. The code you highlighted is the one reporting conflicts (via logging), the actual extraction of the (hopefully only) tunnelPeer happens here:

https://github.com/cilium/cilium/blob/70087b48029c397f412c22b7007299a67a1d90f0/pkg/ipcache/types.go#L155

Due to the asynchronous nature of the API, we cannot report this to the caller when it provides a conflicting value, so we check for conflicts and emit warnings instead.

Reply:
To give a bit more context: For metadata like labels, we can merge them if multiple source provide different labels. But for tunnelIPs, there can only be one IP (since the datapath doesn't support multiple tunnel endpoints), and we therefore assume that this one IP is only coming from a single source.

I we could in theory introduce an equality check, and allow multiple source to specify the tunnel IP, as long as all source agree, but that will due to the eventually consistent nature of our data sources, that will likely lead to temporary states where different sources report a different tunnel IP - thus we currently require that only one source provides a tunnel IP.

Reply:
that makes sense, thanks for the explanation! @gandro